README.md
# WordPress Development Overlay Plugin

WordPress plugin to overlay an image on your page during development.

Once you activate this plugin, on the front-end of your website you'll get a
"Toggle" button.  Clicking the "Toggle" button will hide/reveal the `default.jpg` image from the `/images` directory of this plugin.

See [https://salferrarello.com/wordpress-development-overlay/](https://salferrarello.com/wordpress-development-overlay/)

## Non-WordPress Projects

In a non-WordPress project the JavaScript file,
[js/development-overlay.js](js/development-overlay.js), can be used on its own.
When used on its own, the overlay images are loaded from the relative path
`./overlay/`.

## Configuration

You can replace `default.jpg` with the image you want to overlay on your website during development.

### Advanced Configuration

You can pass further configuration into this plugin via URL parameters.

For example, if the site your are working on is `https://example.test` and you 
want to use a second image without overwriting the first, you can append `?src=another-image.jpg` to load `another-image.jpg` from the images directory of this
plugin.

```
https://example.test?src=another-image.jpg
```

If the image needs to be nudged to the right you can pass in the URL parameter `left`.

e.g.

```
# Add 10px to the left of the image
https://example.test?left=10

# Reduce the space on the left of the image by 20px
https://example.test?left=-20
```

#### Supported URL parameters

The following are supported URL parameters and their default values.

- `src` (`default.jpg`)
- `left` (`0`)
- `top` (`0`)
- `opacity` (`0.5`)
- `display` (`block`)
- `z-index` (`99999`)

## Contributors

- [Sal Ferrarello](https://salferrarello.com) / [@salcode](https://github.com/salcode)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2019-04-04
### Added
- Changelog
- Keyboard support, ESC key toggle overlay
- Add default image path (./overlay/) for projects without WordPress
- Add comment to top of JavaScript file for stand-alone use
- Document use of js/development-overlay.js outside of WordPress

### Removed
- Remove unused readme.txt

## [0.1.0] - 2018-07-14
### Added
- Initial Release


development-overlay.php
<?php
/**
 * Plugin Name:     Development Overlay
 * Plugin URI:      https://salferrarello.com/wordpress-development-overlay/
 * Description:     Overlay an image on your page during development
 * Author:          Sal Ferrarello
 * Author URI:      https://salferrarello.com
 * Text Domain:     development-overlay
 * Domain Path:     /languages
 * Version:         1.0.0
 *
 * @package         development-overlay
 */

// If this file is called directly, abort.
if ( ! defined( 'WPINC' ) ) {
	die;
}

add_action( 'wp_enqueue_scripts', 'fe_development_overlay_enqueue_js' );

/**
 * Enqueue JavaScript
 */
function fe_development_overlay_enqueue_js() {
	wp_enqueue_script(
		'fe-development-overlay',
		plugins_url( 'js/development-overlay.js', __FILE__ ),
		array(),
		'1.0.0',
		true
	);

	wp_localize_script(
		'fe-development-overlay',
		// A name we makeup to avoid our PHP values colliding with others.
		'feDevelopmentOverlay',
		// A key/value array of values to pass from PHP to JavaScript.
		array(
			'path' => plugins_url( 'images', __FILE__ ),
		)
	);
}


js/development-overlay.js
/**
 * Development Overlay
 * Overlay an image on your page during development.
 * https://salferrarello.com/wordpress-development-overlay/
 * by Sal Ferrarello
 * Version: 1.0.0
 */
(function() {

  var css = 'left: 50%; ',
    src = 'default.jpg',

    properties = {
      'display': 'block',
      'opacity': '0.5',
      'margin-left': '0',
      'position': 'absolute',
      'top': '0',
      'transform': 'translateX(-50%)',
      'z-index': '99999'
  },

  propertiesThatNeedPx = [
    'left',
    'margin-left',
    'top'
  ],
  image = document.createElement("img"),
  toggleBtn = document.createElement("button");

  // Update image src if present as URL param.
  if ( getUrlParam( 'src' ) ) {
    src = getUrlParam( 'src' );
  }

  // Add src to image.
  image.src = getImgSrcValue(src);

  // Update CSS.
  Object.keys(properties).forEach(function(key) {

    // Get URL parameter.
    // Use URL param 'left' for our 'margin-left' value.
    var value = getUrlParam('margin-left' === key ? 'left' : key);

    // Use default value from properties as fallback.
    value = value ? value : properties[key];

    // Add 'px' to the value if necessary.
    value = needsPx(key) ? value + 'px' : value;

    // Add to CSS string.
    css += key + ': ' + value + '; ';
  });

  // Apply CSS styles.
  image.style = css;

  // Add image to page.
  document.body.appendChild(image);

  // Add toggle button.
  toggleBtn.className = 'btn btn-warning';
  toggleBtn.style = 'left: 20px; position: fixed; top: 80px; z-index: 99999;'
  toggleBtn.textContent = 'Toggle';

  // Listen for button click.
  toggleBtn.addEventListener('click', toggleImageDisplay);

  // Listen for ESC key (i.e. keyCode 27).
  document.addEventListener('keydown', function(e) {
    if (e.keyCode === 27) { toggleImageDisplay(); }
  });

  // Add button to page.
  document.body.appendChild(toggleBtn);

  function toggleImageDisplay() {
    if ('block' === image.style.display) {
      image.style.display = 'none';
    } else {
      image.style.display = 'block';
    }
  }

  function getUrlParam( key ) {
    var urlParams = new URLSearchParams(
      window.location.href.substr(
        window.location.href.indexOf('?') + 1
      )
    );
    return urlParams.get( key );
  }

  function needsPx(key) {
    return -1 !== propertiesThatNeedPx.indexOf(key);
  }

  function getImgSrcValue(src) {
    if (
      'undefined' !== typeof feDevelopmentOverlay &&
      'undefined' !== typeof feDevelopmentOverlay.path
    ) {
      // Use path provided by feDevelopmentOverlay (via WordPress).
      return feDevelopmentOverlay.path + '/' + src;
    }
    // Use default path.
    return './overlay/' + src;
  }

}());
